File: scripts/dev_tools/codex_native_converter/mapping.py
"""Resolve approved Codex-native target paths for classified artifacts.

Purpose:
    Convert classification records into concrete Codex-native destination paths
    while preserving the approved runtime surfaces defined by the feature scope.

Usage:
    The converter engine calls ``plan_target_paths`` after classification and
    before validation.

Flow:
    The planner chooses a target path from the record's target role, applies
    naming normalization, and returns a copy of the mapping record with the
    planned target path.

Invariants / Constraints:
    Only approved Codex-native surfaces are emitted. Repository-convention
    prompt output remains disabled unless the caller explicitly enables it.

Side Effects:
    None.
"""

from __future__ import annotations

from dataclasses import replace
from pathlib import PurePosixPath

from scripts.dev_tools.codex_native_converter.models import (
    ConversionClass,
    MappingRecord,
    SourceEcosystem,
    SourceKind,
    TargetRole,
)


def _normalized_name(source_path: str) -> str:
    """Normalize one source path into a stable target-friendly base name.

    Purpose:
        Produce consistent skill, agent, hook, and prompt names from mixed
        source filenames and folder structures.

    Args:
        source_path (str): Source-root-relative path for the artifact.

    Returns:
        str: A kebab-case friendly base name without ecosystem-specific suffixes.

    Raises:
        None.

    Side Effects:
        None.
    """

    source_name = PurePosixPath(source_path).name
    normalized_name = source_name
    for suffix in (".instructions.md", ".agent.md", ".prompt.md", ".md", ".json"):
        if normalized_name.endswith(suffix):
            normalized_name = normalized_name[: -len(suffix)]
            break
    return normalized_name.replace("_", "-")


def _planned_skill_name(mapping_record: MappingRecord) -> str:
    """Derive the target skill name for one mapped skill-like artifact.

    Purpose:
        Preserve the reusable skill folder identity for `*/skills/<name>/SKILL.md`
        while keeping filename-based naming for path-scoped instruction files.

    Args:
        mapping_record (MappingRecord): Skill-like mapping record whose target
            folder name must be planned.

    Returns:
        str: The normalized skill folder name for the target path.

    Raises:
        None.

    Side Effects:
        None.
    """

    source_path = PurePosixPath(mapping_record.source_path)
    if (
        mapping_record.source_kind is SourceKind.REUSABLE_SKILL
        and source_path.name == "SKILL.md"
        and source_path.parent.name
    ):
        return source_path.parent.name.replace("_", "-")
    return _normalized_name(mapping_record.source_path)


def _planned_hook_name(mapping_record: MappingRecord) -> str:
    """Derive the target hook name without carrying source script extensions."""

    source_name = PurePosixPath(mapping_record.source_path).name
    if source_name.endswith((".ps1", ".py")):
        normalized_name = source_name.rsplit(".", 1)[0]
        return normalized_name.replace("_", "-")
    return _normalized_name(mapping_record.source_path)


def plan_target_paths(
    mapping_record: MappingRecord,
    *,
    enable_repo_prompts: bool,
) -> MappingRecord:
    """Resolve the approved Codex-native target path for one mapping record.

    Purpose:
        Attach a concrete target path to a classified mapping record when the
        approved v1 target surface is known.

    Args:
        mapping_record (MappingRecord): Classified mapping record whose target
            path is still unresolved.
        enable_repo_prompts (bool): Whether repository-convention prompt output
            is enabled for this run.

    Returns:
        MappingRecord: A copy of the input record with ``target_path`` resolved
        when the artifact has an approved destination.

    Raises:
        None.

    Side Effects:
        None.
    """

    if mapping_record.target_role is TargetRole.STANDING_GUIDANCE:
        return replace(mapping_record, target_path="AGENTS.md")

    if mapping_record.target_role is TargetRole.SHARED_SKILL:
        skill_name = _planned_skill_name(mapping_record)
        return replace(
            mapping_record,
            target_path=f".agents/skills/{skill_name}/SKILL.md",
        )

    if mapping_record.target_role is TargetRole.SUBAGENT:
        agent_name = _normalized_name(mapping_record.source_path)
        return replace(
            mapping_record,
            target_path=f".codex/agents/{agent_name}.toml",
        )

    if mapping_record.target_role is TargetRole.MCP_CONFIG:
        return replace(mapping_record, target_path=".codex/config.toml")

    if mapping_record.target_role is TargetRole.HOOK:
        hook_name = _planned_hook_name(mapping_record)
        return replace(
            mapping_record,
            target_path=f".codex/hooks/{hook_name}.ps1",
        )

    if mapping_record.target_role is TargetRole.APPROVAL_RULE:
        rule_name = _normalized_name(mapping_record.source_path)
        return replace(
            mapping_record,
            target_path=f".codex/rules/{rule_name}.rules",
        )

    if mapping_record.target_role is TargetRole.LAUNCHER:
        if enable_repo_prompts:
            prompt_name = _normalized_name(mapping_record.source_path)
            return replace(
                mapping_record,
                target_path=f".codex/prompts/{prompt_name}.md",
            )
        return replace(
            mapping_record,
            conversion_class=ConversionClass.UNSUPPORTED,
            target_role=TargetRole.UNSUPPORTED,
            target_path=None,
            notes=(
                *mapping_record.notes,
                "Repository-convention prompt output is disabled for this run.",
            ),
            is_required=False,
        )

    return replace(mapping_record, target_path=None)


def plan_section_target_path(
    source_path: str,
    *,
    source_ecosystem: SourceEcosystem,
    source_kind: SourceKind,
    target_role: TargetRole,
    enable_repo_prompts: bool,
) -> str | None:
    """Resolve a native target path for one section-level planned emission.

    Purpose:
        Reuse the file-level path-planning rules for section-level prompt and
        mixed-artifact emissions without duplicating naming logic.

    Args:
        source_path (str): Source-root-relative artifact path.
        source_kind (SourceKind): Source kind associated with the section.
        target_role (TargetRole): Native role selected for the section.
        enable_repo_prompts (bool): Whether repository-convention prompt output
            is enabled for the current run.

    Returns:
        str | None: Planned target path when the role has an approved native
            destination, otherwise None.

    Raises:
        None.

    Side Effects:
        None.
    """

    return plan_target_paths(
        MappingRecord(
            source_path=source_path,
            source_ecosystem=source_ecosystem,
            source_kind=source_kind,
            conversion_class=ConversionClass.DECOMPOSED,
            target_role=target_role,
            target_path=None,
        ),
        enable_repo_prompts=enable_repo_prompts,
    ).target_path

File: extensions/drm-copilot/src/lib/codex-native-converter/mapping.ts
/**
 * Resolve approved Codex-native target paths for classified artifacts.
 *
 * Purpose:
 *     Convert classification records into concrete Codex-native destination
 *     paths while preserving the approved runtime surfaces. Ported from
 *     `mapping.py`; pure logic, no I/O.
 *
 * Invariants:
 *     Only approved Codex-native surfaces are emitted. Repository-convention
 *     prompt output remains disabled unless the caller explicitly enables it.
 */

import {
  ConversionClass,
  type MappingRecord,
  type SourceEcosystem,
  SourceKind,
  TargetRole,
} from "./models";

/**
 * Return the final path segment (basename) of a POSIX path.
 *
 * @param posixPath POSIX path.
 * @returns The basename.
 */
function posixName(posixPath: string): string {
  const trimmed = posixPath.replace(/\/+$/, "");
  const index = trimmed.lastIndexOf("/");
  return index >= 0 ? trimmed.slice(index + 1) : trimmed;
}

/**
 * Return the parent folder name of a POSIX path, or an empty string when there
 * is no parent segment.
 *
 * @param posixPath POSIX path.
 * @returns The parent directory name.
 */
function posixParentName(posixPath: string): string {
  const trimmed = posixPath.replace(/\/+$/, "");
  const index = trimmed.lastIndexOf("/");
  if (index < 0) {
    return "";
  }
  return posixName(trimmed.slice(0, index));
}

/**
 * Normalize one source path into a stable target-friendly base name.
 *
 * Mirrors `_normalized_name`: strips known ecosystem suffixes (first match
 * only) and replaces underscores with hyphens.
 *
 * @param sourcePath Source-root-relative path for the artifact.
 * @returns A kebab-case friendly base name.
 */
function normalizedName(sourcePath: string): string {
  const sourceName = posixName(sourcePath);
  let normalized = sourceName;
  // Strip the first matching known suffix.
  for (const suffix of [
    ".instructions.md",
    ".agent.md",
    ".prompt.md",
    ".md",
    ".json",
  ]) {
    if (normalized.endsWith(suffix)) {
      normalized = normalized.slice(0, normalized.length - suffix.length);
      break;
    }
  }
  return normalized.replace(/_/g, "-");
}

/**
 * Derive the target skill name for one mapped skill-like artifact.
 *
 * Mirrors `_planned_skill_name`: a reusable `SKILL.md` keeps its parent folder
 * identity; otherwise the filename-based name is used.
 *
 * @param mappingRecord Skill-like mapping record.
 * @returns The normalized skill folder name.
 */
function plannedSkillName(mappingRecord: MappingRecord): string {
  const name = posixName(mappingRecord.sourcePath);
  const parentName = posixParentName(mappingRecord.sourcePath);
  if (
    mappingRecord.sourceKind === SourceKind.REUSABLE_SKILL &&
    name === "SKILL.md" &&
    parentName !== ""
  ) {
    return parentName.replace(/_/g, "-");
  }
  return normalizedName(mappingRecord.sourcePath);
}

/**
 * Derive the target hook name without carrying source script extensions.
 *
 * Mirrors `_planned_hook_name`.
 *
 * @param mappingRecord Hook mapping record.
 * @returns The normalized hook name.
 */
function plannedHookName(mappingRecord: MappingRecord): string {
  const sourceName = posixName(mappingRecord.sourcePath);
  if (sourceName.endsWith(".ps1") || sourceName.endsWith(".py")) {
    const normalized = sourceName.slice(0, sourceName.lastIndexOf("."));
    return normalized.replace(/_/g, "-");
  }
  return normalizedName(mappingRecord.sourcePath);
}

/**
 * Resolve the approved Codex-native target path for one mapping record.
 *
 * Mirrors `plan_target_paths`, preserving every target-role branch, the
 * `enable_repo_prompts` `.codex/prompts` branch, the `AGENTS.md` standing
 * target, the unsupported-launcher fallthrough, and the null target for
 * unmapped roles.
 *
 * @param mappingRecord Classified mapping record whose target path is unresolved.
 * @param options Run options (`enableRepoPrompts`).
 * @returns A copy of the input record with `targetPath` resolved when known.
 */
export function planTargetPaths(
  mappingRecord: MappingRecord,
  options: { readonly enableRepoPrompts: boolean },
): MappingRecord {
  const { enableRepoPrompts } = options;

  if (mappingRecord.targetRole === TargetRole.STANDING_GUIDANCE) {
    return { ...mappingRecord, targetPath: "AGENTS.md" };
  }

  if (mappingRecord.targetRole === TargetRole.SHARED_SKILL) {
    const skillName = plannedSkillName(mappingRecord);
    return {
      ...mappingRecord,
      targetPath: `.agents/skills/${skillName}/SKILL.md`,
    };
  }

  if (mappingRecord.targetRole === TargetRole.SUBAGENT) {
    const agentName = normalizedName(mappingRecord.sourcePath);
    return { ...mappingRecord, targetPath: `.codex/agents/${agentName}.toml` };
  }

  if (mappingRecord.targetRole === TargetRole.MCP_CONFIG) {
    return { ...mappingRecord, targetPath: ".codex/config.toml" };
  }

  if (mappingRecord.targetRole === TargetRole.HOOK) {
    const hookName = plannedHookName(mappingRecord);
    return { ...mappingRecord, targetPath: `.codex/hooks/${hookName}.ps1` };
  }

  if (mappingRecord.targetRole === TargetRole.APPROVAL_RULE) {
    const ruleName = normalizedName(mappingRecord.sourcePath);
    return { ...mappingRecord, targetPath: `.codex/rules/${ruleName}.rules` };
  }

  if (mappingRecord.targetRole === TargetRole.LAUNCHER) {
    if (enableRepoPrompts) {
      const promptName = normalizedName(mappingRecord.sourcePath);
      return {
        ...mappingRecord,
        targetPath: `.codex/prompts/${promptName}.md`,
      };
    }
    // Repo-convention prompts disabled: demote to unsupported.
    return {
      ...mappingRecord,
      conversionClass: ConversionClass.UNSUPPORTED,
      targetRole: TargetRole.UNSUPPORTED,
      targetPath: null,
      notes: [
        ...mappingRecord.notes,
        "Repository-convention prompt output is disabled for this run.",
      ],
      isRequired: false,
    };
  }

  return { ...mappingRecord, targetPath: null };
}

/**
 * Resolve a native target path for one section-level planned emission.
 *
 * Mirrors `plan_section_target_path`: reuses the file-level planner for a
 * synthetic decomposed record and returns the resolved target path.
 *
 * @param sourcePath Source-root-relative artifact path.
 * @param options Section ecosystem/kind/role plus `enableRepoPrompts`.
 * @returns The planned target path when the role has a destination, else null.
 */
export function planSectionTargetPath(
  sourcePath: string,
  options: {
    readonly sourceEcosystem: SourceEcosystem;
    readonly sourceKind: SourceKind;
    readonly targetRole: TargetRole;
    readonly enableRepoPrompts: boolean;
  },
): string | null {
  return planTargetPaths(
    {
      sourcePath,
      sourceEcosystem: options.sourceEcosystem,
      sourceKind: options.sourceKind,
      conversionClass: ConversionClass.DECOMPOSED,
      targetRole: options.targetRole,
      targetPath: null,
      notes: [],
      isRequired: true,
    },
    { enableRepoPrompts: options.enableRepoPrompts },
  ).targetPath;
}
